### ISSUE TITLE
Change input scheme for ParticleFactory

### ISSUE THREAD

**Issue by greenwald** [yap/YAP#69]
Let's replace `ParticleFactory::readPDT(const std::string fname)` with something using `istream_iterator`.

A rough first idea:

* implement `ParticleFactory::insert(iterator pos, ParticleTableEntry)`
* implement `istream_iterator` to read a full line of a file.
* implement a function to convert that full line to a `ParticleTableEntry` for the PDT format.

We would like to fill the table with:
```cpp
template <class line_reader>
void fillTable(istream_iterator first, istream_iterator last, line_reader LR)
{
  std::transform(first, last, std::inserter(ParticleTable_, ParticleTable_.end()), LR);
}
```

`line_read` should be an object that has a functor such that the object can be called on a `istream_iterator`, and produce a `ParticleTableEntry`

**Comment by greenwald** [comment]
```cpp
template <class Iter>
ParticleFactory::ParticleFactory(Iter first, Iter last)
{
    std::copy(first, last, std::inserter(ParticleTable_, ParticleTable_.end()));
}

auto PF = ParticleFactory(yap::XXXIterator(file), yap::XXXIterator::end());
```

In ParticeFactory add
```cpp
using ParticleTableMap = std::map<int, ParticleTableEntry>;
```

So: the iterators above must dereference to a `ParticleTableMap::value_type` and the increment operator must increment some internal structure of the iterator to state from which you can construct or read the next `ParticleTableEntry`.

For example, for PDL:
* The `PDLIterator` would store an `istream_iterator` to a file inside it.
* Dereferencing constructs a `ParticleTableEntry` from the point of the internal `istream_iterator`
* Incrementing increments until the point starting the next entry (a line starting with `"add"`) or failing that, sets the iterator to the end.

You can also make a helper function, inside the file, say, for `PDLIterator`:
```cpp
ParticleFactory read_pdl_file(std::string filename)
{ return ParticleFactory(PDLIterator(fstream(filename)), PDLIterator::end()); }
```

**Comment by pdigiglio** [comment]
__Edit:__
 - [ ] `ParticleFactory`'s constructor should call `std::copy` through another method, let's say `importTable`;
 - [ ] Implement `operator+=` for `ParticleFactory`.

**Comment by greenwald** [comment]
Closed via #76
